File: app/src/main/java/awais/instagrabber/fragments/directmessages/DirectMessageSettingsFragment.java
package awais.instagrabber.fragments.directmessages;

import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Pair;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelStoreOwner;
import androidx.navigation.NavBackStackEntry;
import androidx.navigation.NavController;
import androidx.navigation.NavDestination;
import androidx.navigation.NavDirections;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.Set;

import awais.instagrabber.R;
import awais.instagrabber.adapters.DirectUsersAdapter;
import awais.instagrabber.customviews.helpers.TextWatcherAdapter;
import awais.instagrabber.databinding.FragmentDirectMessagesSettingsBinding;
import awais.instagrabber.dialogs.MultiOptionDialogFragment;
import awais.instagrabber.dialogs.MultiOptionDialogFragment.Option;
import awais.instagrabber.models.Resource;
import awais.instagrabber.repositories.responses.directmessages.DirectThread;
import awais.instagrabber.repositories.responses.directmessages.DirectUser;
import awais.instagrabber.viewmodels.DirectInboxViewModel;
import awais.instagrabber.viewmodels.DirectSettingsViewModel;

public class DirectMessageSettingsFragment extends Fragment {
    private static final String TAG = DirectMessageSettingsFragment.class.getSimpleName();

    private FragmentDirectMessagesSettingsBinding binding;
    private DirectSettingsViewModel viewModel;
    private DirectUsersAdapter usersAdapter;
    private List<Option<String>> options;

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final NavController navController = NavHostFragment.findNavController(this);
        final ViewModelStoreOwner viewModelStoreOwner = navController.getViewModelStoreOwner(R.id.direct_messages_nav_graph);
        final DirectInboxViewModel inboxViewModel = new ViewModelProvider(viewModelStoreOwner).get(DirectInboxViewModel.class);
        final List<DirectThread> threads = inboxViewModel.getThreads().getValue();
        final Bundle arguments = getArguments();
        if (arguments == null) {
            navController.navigateUp();
            return;
        }
        final DirectMessageSettingsFragmentArgs fragmentArgs = DirectMessageSettingsFragmentArgs.fromBundle(arguments);
        final String threadId = fragmentArgs.getThreadId();
        final Optional<DirectThread> first = threads != null ? threads.stream()
                                                                      .filter(thread -> thread.getThreadId().equals(threadId))
                                                                      .findFirst()
                                                             : Optional.empty();
        if (!first.isPresent()) {
            navController.navigateUp();
            return;
        }
        viewModel = new ViewModelProvider(this).get(DirectSettingsViewModel.class);
        viewModel.setViewer(inboxViewModel.getViewer());
        viewModel.setThread(first.get());
        // basicClickListener = v -> {
        //     final Object tag = v.getTag();
        //     if (tag instanceof ProfileModel) {
        //         ProfileModel model = (ProfileModel) tag;
        //         final Bundle bundle = new Bundle();
        //         bundle.putString("username", "@" + model.getUsername());
        //         NavHostFragment.findNavController(this).navigate(R.id.action_global_profileFragment, bundle);
        //     }
        // };
        //
        // clickListener = v -> {
        //     final Object tag = v.getTag();
        //     if (tag instanceof ProfileModel) {
        //         ProfileModel model = (ProfileModel) tag;
        //         final Context context = getContext();
        //         if (context == null) return;
        //         final ArrayAdapter<String> adapter = new ArrayAdapter<>(context, android.R.layout.simple_list_item_1, new String[]{
        //                 getString(R.string.open_profile),
        //                 getString(R.string.dms_action_kick),
        //         });
        //         final DialogInterface.OnClickListener clickListener = (d, w) -> {
        //             if (w == 0) {
        //                 final Bundle bundle = new Bundle();
        //                 bundle.putString("username", "@" + model.getUsername());
        //                 NavHostFragment.findNavController(this).navigate(R.id.action_global_profileFragment, bundle);
        //             } else if (w == 1) {
        //                 new ChangeSettings(titleText.getText().toString()).execute("remove_users", model.getId());
        //                 onRefresh();
        //             }
        //         };
        //         new AlertDialog.Builder(context)
        //                 .setAdapter(adapter, clickListener)
        //                 .show();
        //     }
        // };
    }

    @NonNull
    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater,
                             final ViewGroup container,
                             final Bundle savedInstanceState) {
        binding = FragmentDirectMessagesSettingsBinding.inflate(inflater, container, false);
        // final String threadId = DirectMessageSettingsFragmentArgs.fromBundle(getArguments()).getThreadId();
        // threadTitle = DirectMessageSettingsFragmentArgs.fromBundle(getArguments()).getTitle();
        // binding.swipeRefreshLayout.setEnabled(false);

        // final ActionBar actionBar = fragmentActivity.getSupportActionBar();
        // if (actionBar != null) {
        //     actionBar.setTitle(threadTitle);
        // }

        // titleSend.setOnClickListener(v -> new ChangeSettings(titleText.getText().toString()).execute("update_title"));

        // binding.titleText.addTextChangedListener(new TextWatcherAdapter() {
        //     @Override
        //     public void onTextChanged(CharSequence s, int start, int before, int count) {
        //         binding.titleSend.setVisibility(s.toString().equals(threadTitle) ? View.GONE : View.VISIBLE);
        //     }
        // });

        // final AppCompatButton btnLeave = binding.btnLeave;
        // btnLeave.setOnClickListener(v -> new AlertDialog.Builder(context)
        //         .setTitle(R.string.dms_action_leave_question)
        //         .setPositiveButton(R.string.yes, (x, y) -> new ChangeSettings(titleText.getText().toString()).execute("leave"))
        //         .setNegativeButton(R.string.no, null)
        //         .show());

        // currentlyRunning = new DirectMessageInboxThreadFetcher(threadId, null, null, fetchListener).execute();
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState) {
        init();
        setupObservers();
    }

    private void setupObservers() {
        viewModel.getUsers().observe(getViewLifecycleOwner(), users -> {
            if (usersAdapter == null) return;
            usersAdapter.submitUsers(users.first, users.second);
        });
        viewModel.getTitle().observe(getViewLifecycleOwner(), title -> binding.titleEdit.setText(title));
        viewModel.getAdminUserIds().observe(getViewLifecycleOwner(), adminUserIds -> {
            if (usersAdapter == null) return;
            usersAdapter.setAdminUserIds(adminUserIds);
        });
        final NavController navController = NavHostFragment.findNavController(this);
        final NavBackStackEntry backStackEntry = navController.getCurrentBackStackEntry();
        if (backStackEntry != null) {
            final MutableLiveData<Object> resultLiveData = backStackEntry.getSavedStateHandle().getLiveData("result");
            resultLiveData.observe(getViewLifecycleOwner(), result -> {
                LiveData<Resource<Object>> detailsChangeResourceLiveData = null;
                if ((result instanceof DirectUser)) {
                    // Log.d(TAG, "result: " + result);
                    detailsChangeResourceLiveData = viewModel.addMembers(Collections.singleton((DirectUser) result));
                } else if ((result instanceof Set)) {
                    try {
                        // Log.d(TAG, "result: " + result);
                        //noinspection unchecked
                        detailsChangeResourceLiveData = viewModel.addMembers((Set<DirectUser>) result);
                    } catch (Exception e) {
                        Log.e(TAG, "search users result: ", e);
                    }
                }
                if (detailsChangeResourceLiveData != null) {
                    observeDetailsChange(detailsChangeResourceLiveData);
                }
            });
        }
    }

    private void init() {
        setupSettings();
        setupMembers();
    }

    private void setupSettings() {
        binding.groupSettings.setVisibility(viewModel.isGroup() ? View.VISIBLE : View.GONE);
        if (!viewModel.isGroup()) return;
        binding.titleEdit.addTextChangedListener(new TextWatcherAdapter() {
            @Override
            public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {
                if (s.toString().trim().equals(viewModel.getTitle().getValue())) {
                    binding.titleEditInputLayout.setSuffixText(null);
                    return;
                }
                binding.titleEditInputLayout.setSuffixText(getString(R.string.save));
            }
        });
        binding.titleEditInputLayout.getSuffixTextView().setOnClickListener(v -> {
            final Editable text = binding.titleEdit.getText();
            if (text == null) return;
            final String newTitle = text.toString().trim();
            if (newTitle.equals(viewModel.getTitle().getValue())) return;
            observeDetailsChange(viewModel.updateTitle(newTitle));
        });
        binding.addMembers.setOnClickListener(v -> {
            if (!isAdded()) return;
            final NavController navController = NavHostFragment.findNavController(this);
            final NavDestination currentDestination = navController.getCurrentDestination();
            if (currentDestination == null) return;
            if (currentDestination.getId() != R.id.directMessagesSettingsFragment) return;
            final Pair<List<DirectUser>, List<DirectUser>> users = viewModel.getUsers().getValue();
            final long[] currentUserIds;
            if (users != null && users.first != null) {
                final List<DirectUser> currentMembers = users.first;
                currentUserIds = currentMembers.stream()
                                               .mapToLong(DirectUser::getPk)
                                               .sorted()
                                               .toArray();
            } else {
                currentUserIds = new long[0];
            }
            final NavDirections directions = DirectMessageSettingsFragmentDirections.actionGlobalUserSearch(
                    true,
                    "Add users",
                    "Add",
                    currentUserIds
            );
            navController.navigate(directions);
        });
    }

    private void setupMembers() {
        final Context context = getContext();
        if (context == null) return;
        binding.users.setLayoutManager(new LinearLayoutManager(context));
        final DirectUser inviter = viewModel.getThread().getInviter();
        usersAdapter = new DirectUsersAdapter(
                inviter != null ? inviter.getPk() : -1,
                (position, user, selected) -> {
                    // navigate to profile
                },
                (position, user) -> {
                    final ArrayList<Option<String>> options = viewModel.createUserOptions(user);
                    if (options == null || options.isEmpty()) return true;
                    final MultiOptionDialogFragment<String> fragment = MultiOptionDialogFragment.newInstance(-1, options);
                    fragment.setSingleCallback(new MultiOptionDialogFragment.MultiOptionDialogSingleCallback<String>() {
                        @Override
                        public void onSelect(final String action) {
                            if (action == null) return;
                            observeDetailsChange(viewModel.doAction(user, action));
                        }

                        @Override
                        public void onCancel() {}
                    });
                    final FragmentManager fragmentManager = getChildFragmentManager();
                    fragment.show(fragmentManager, "actions");
                    return true;
                }
        );
        binding.users.setAdapter(usersAdapter);
    }

    private void observeDetailsChange(@NonNull final LiveData<Resource<Object>> detailsChangeResourceLiveData) {
        detailsChangeResourceLiveData.observe(getViewLifecycleOwner(), resource -> {
            if (resource == null) return;
            switch (resource.status) {
                case SUCCESS:
                case LOADING:
                    break;
                case ERROR:
                    if (resource.message != null) {
                        Snackbar.make(binding.getRoot(), resource.message, Snackbar.LENGTH_LONG).show();
                    }
                    break;
            }
        });
    }

    // class ChangeSettings extends AsyncTask<String, Void, Void> {
    //     String action, argument;
    //     boolean ok = false;
    //     private final String text;
    //
    //     public ChangeSettings(final String text) {
    //         this.text = text;
    //     }
    //
    //     protected Void doInBackground(String... rawAction) {
    //         action = rawAction[0];
    //         if (rawAction.length == 2) argument = rawAction[1];
    //         final String url = "https://i.instagram.com/api/v1/direct_v2/threads/" + threadId + "/" + action + "/";
    //         try {
    //             String urlParameters = "_csrftoken=" + cookie.split("csrftoken=")[1].split(";")[0]
    //                     + "&_uuid=" + Utils.settingsHelper.getString(Constants.DEVICE_UUID);
    //             if (action.equals("update_title")) {
    //                 urlParameters += "&title=" + URLEncoder.encode(text, "UTF-8")
    //                                                        .replaceAll("\\+", "%20").replaceAll("%21", "!").replaceAll("%27", "'")
    //                                                        .replaceAll("%28", "(").replaceAll("%29", ")").replaceAll("%7E", "~");
    //             } else if (action.startsWith("remove_users"))
    //                 urlParameters += ("&user_ids=[" + argument + "]");
    //             final HttpURLConnection urlConnection = (HttpURLConnection) new URL(url).openConnection();
    //             urlConnection.setRequestMethod("POST");
    //             urlConnection.setUseCaches(false);
    //             urlConnection.setRequestProperty("User-Agent", Constants.I_USER_AGENT);
    //             urlConnection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
    //             urlConnection.setRequestProperty("Content-Length", Integer.toString(urlParameters.getBytes().length));
    //             urlConnection.setDoOutput(true);
    //             DataOutputStream wr = new DataOutputStream(urlConnection.getOutputStream());
    //             wr.writeBytes(urlParameters);
    //             wr.flush();
    //             wr.close();
    //             urlConnection.connect();
    //             if (urlConnection.getResponseCode() == HttpURLConnection.HTTP_OK) {
    //                 ok = true;
    //             }
    //             urlConnection.disconnect();
    //         } catch (Throwable ex) {
    //             Log.e("austin_debug", "unsend: " + ex);
    //         }
    //         return null;
    //     }
    //
    //     @Override
    //     protected void onPostExecute(Void result) {
    //         final Context context = getContext();
    //         if (context == null) return;
    //         if (ok) {
    //             Toast.makeText(context, R.string.dms_action_success, Toast.LENGTH_SHORT).show();
    //             if (action.equals("update_title")) {
    //                 threadTitle = titleText.getText().toString();
    //                 titleSend.setVisibility(View.GONE);
    //                 titleText.clearFocus();
    //                 DirectMessageThreadFragment.hasSentSomething = true;
    //             } else if (action.equals("leave")) {
    //                 context.sendBroadcast(new Intent(DMRefreshBroadcastReceiver.ACTION_REFRESH_DM));
    //                 NavHostFragment.findNavController(DirectMessageSettingsFragment.this).popBackStack(R.id.directMessagesInboxFragment, false);
    //             } else {
    //                 DirectMessageThreadFragment.hasSentSomething = true;
    //             }
    //         } else Toast.makeText(context, R.string.downloader_unknown_error, Toast.LENGTH_SHORT).show();
    //     }
    // }
}

File: app/src/main/java/awais/instagrabber/fragments/directmessages/DirectMessageSettingsFragment.kt
package awais.instagrabber.fragments.directmessages

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.LinearLayoutManager
import awais.instagrabber.ProfileNavGraphDirections
import awais.instagrabber.R
import awais.instagrabber.activities.MainActivity
import awais.instagrabber.adapters.DirectPendingUsersAdapter
import awais.instagrabber.adapters.DirectPendingUsersAdapter.PendingUser
import awais.instagrabber.adapters.DirectPendingUsersAdapter.PendingUserCallback
import awais.instagrabber.adapters.DirectUsersAdapter
import awais.instagrabber.customviews.helpers.TextWatcherAdapter
import awais.instagrabber.databinding.FragmentDirectMessagesSettingsBinding
import awais.instagrabber.dialogs.ConfirmDialogFragment
import awais.instagrabber.dialogs.ConfirmDialogFragment.ConfirmDialogFragmentCallback
import awais.instagrabber.dialogs.MultiOptionDialogFragment
import awais.instagrabber.dialogs.MultiOptionDialogFragment.MultiOptionDialogSingleCallback
import awais.instagrabber.fragments.UserSearchFragment
import awais.instagrabber.fragments.UserSearchFragmentDirections
import awais.instagrabber.models.Resource
import awais.instagrabber.repositories.responses.User
import awais.instagrabber.repositories.responses.directmessages.DirectThreadParticipantRequestsResponse
import awais.instagrabber.repositories.responses.directmessages.RankedRecipient
import awais.instagrabber.utils.TextUtils.isEmpty
import awais.instagrabber.utils.Utils
import awais.instagrabber.utils.extensions.TAG
import awais.instagrabber.viewmodels.AppStateViewModel
import awais.instagrabber.viewmodels.DirectSettingsViewModel
import awais.instagrabber.viewmodels.factories.DirectSettingsViewModelFactory
import com.google.android.material.snackbar.Snackbar
import java.util.*

class DirectMessageSettingsFragment : Fragment(), ConfirmDialogFragmentCallback {
    private lateinit var viewModel: DirectSettingsViewModel
    private lateinit var binding: FragmentDirectMessagesSettingsBinding

    private var usersAdapter: DirectUsersAdapter? = null
    private var isPendingRequestsSetupDone = false
    private var pendingUsersAdapter: DirectPendingUsersAdapter? = null
    private var approvalRequiredUsers: Set<User>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val arguments = arguments ?: return
        val args = DirectMessageSettingsFragmentArgs.fromBundle(arguments)
        val fragmentActivity = requireActivity() as MainActivity
        val appStateViewModel: AppStateViewModel by activityViewModels()
        val currentUser = appStateViewModel.currentUser ?: return
        val viewModelFactory = DirectSettingsViewModelFactory(
            fragmentActivity.application,
            args.threadId,
            args.pending,
            currentUser
        )
        viewModel = ViewModelProvider(this, viewModelFactory).get(DirectSettingsViewModel::class.java)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        binding = FragmentDirectMessagesSettingsBinding.inflate(inflater, container, false)
        // currentlyRunning = new DirectMessageInboxThreadFetcher(threadId, null, null, fetchListener).execute();
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        init()
        setupObservers()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        isPendingRequestsSetupDone = false
    }

    private fun setupObservers() {
        viewModel.inputMode.observe(viewLifecycleOwner, { inputMode: Int? ->
            if (inputMode == null || inputMode == 0) return@observe
            if (inputMode == 1) {
                binding.groupSettings.visibility = View.GONE
                binding.pendingMembersGroup.visibility = View.GONE
                binding.approvalRequired.visibility = View.GONE
                binding.approvalRequiredLabel.visibility = View.GONE
                binding.muteMessagesLabel.visibility = View.GONE
                binding.muteMessages.visibility = View.GONE
            }
        })
        // Need to observe, so that getValue is correct
        viewModel.getUsers().observe(viewLifecycleOwner, { })
        viewModel.getLeftUsers().observe(viewLifecycleOwner, { })
        viewModel.getUsersAndLeftUsers().observe(viewLifecycleOwner, { usersAdapter?.submitUsers(it.first, it.second) })
        viewModel.getTitle().observe(viewLifecycleOwner, { binding.titleEdit.setText(it) })
        viewModel.getAdminUserIds().observe(viewLifecycleOwner, { usersAdapter?.setAdminUserIds(it) })
        viewModel.isMuted().observe(viewLifecycleOwner, { binding.muteMessages.isChecked = it })
        viewModel.isPending().observe(viewLifecycleOwner, { binding.muteMessages.visibility = if (it) View.GONE else View.VISIBLE })
        viewModel.isViewerAdmin().observe(viewLifecycleOwner, { setApprovalRelatedUI(it) })
        viewModel.getApprovalRequiredToJoin().observe(viewLifecycleOwner, { binding.approvalRequired.isChecked = it })
        viewModel.getPendingRequests().observe(viewLifecycleOwner, { setPendingRequests(it) })
        viewModel.isGroup().observe(viewLifecycleOwner, { isGroup: Boolean -> setupSettings(isGroup) })
        val navController = NavHostFragment.findNavController(this)
        val backStackEntry = navController.currentBackStackEntry
        if (backStackEntry != null) {
            val resultLiveData = backStackEntry.savedStateHandle.getLiveData<Any>("result")
            resultLiveData.observe(viewLifecycleOwner, { result: Any? ->
                if (result == null) return@observe
                if (result is RankedRecipient) {
                    val user = getUser(result)
                    // Log.d(TAG, "result: " + user);
                    if (user != null) {
                        addMembers(setOf(user))
                    }
                } else if (result is Set<*>) {
                    try {
                        @Suppress("UNCHECKED_CAST") val recipients = result as Set<RankedRecipient>
                        val users: Set<User> = recipients.asSequence()
                            .filterNotNull()
                            .map { getUser(it) }
                            .filterNotNull()
                            .toSet()
                        // Log.d(TAG, "result: " + users);
                        addMembers(users)
                    } catch (e: Exception) {
                        Log.e(TAG, "search users result: ", e)
                        Snackbar.make(binding.root, e.message ?: "", Snackbar.LENGTH_LONG).show()
                    }
                }
            })
        }
    }

    private fun addMembers(users: Set<User>) {
        val approvalRequired = viewModel.getApprovalRequiredToJoin().value
        var isViewerAdmin = viewModel.isViewerAdmin().value
        if (isViewerAdmin == null) {
            isViewerAdmin = false
        }
        if (!isViewerAdmin && approvalRequired != null && approvalRequired) {
            approvalRequiredUsers = users
            val confirmDialogFragment = ConfirmDialogFragment.newInstance(
                APPROVAL_REQUIRED_REQUEST_CODE,
                R.string.admin_approval_required,
                R.string.admin_approval_required_description,
                R.string.ok,
                R.string.cancel,
                0
            )
            confirmDialogFragment.show(childFragmentManager, "approval_required_dialog")
            return
        }
        val detailsChangeResourceLiveData = viewModel.addMembers(users)
        observeDetailsChange(detailsChangeResourceLiveData)
    }

    private fun getUser(recipient: RankedRecipient): User? {
        var user: User? = null
        if (recipient.user != null) {
            user = recipient.user
        } else if (recipient.thread != null && !recipient.thread.isGroup) {
            user = recipient.thread.users?.get(0)
        }
        return user
    }

    private fun init() {
        // setupSettings();
        setupMembers()
    }

    private fun setupSettings(isGroup: Boolean) {
        binding.groupSettings.visibility = if (isGroup) View.VISIBLE else View.GONE
        binding.muteMessagesLabel.setOnClickListener { binding.muteMessages.toggle() }
        binding.muteMessages.setOnCheckedChangeListener { buttonView: CompoundButton, isChecked: Boolean ->
            val resourceLiveData = if (isChecked) viewModel.mute() else viewModel.unmute()
            handleSwitchChangeResource(resourceLiveData, buttonView)
        }
        if (!isGroup) return
        binding.titleEdit.addTextChangedListener(object : TextWatcherAdapter() {
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (s.toString().trim { it <= ' ' } == viewModel.getTitle().value) {
                    binding.titleEditInputLayout.suffixText = null
                    return
                }
                binding.titleEditInputLayout.suffixText = getString(R.string.save)
            }
        })
        binding.titleEditInputLayout.suffixTextView.setOnClickListener {
            val text = binding.titleEdit.text ?: return@setOnClickListener
            val newTitle = text.toString().trim { it <= ' ' }
            if (newTitle == viewModel.getTitle().value) return@setOnClickListener
            observeDetailsChange(viewModel.updateTitle(newTitle))
        }
        binding.addMembers.setOnClickListener {
            if (!isAdded) return@setOnClickListener
            val navController = NavHostFragment.findNavController(this)
            val currentDestination = navController.currentDestination ?: return@setOnClickListener
            if (currentDestination.id != R.id.directMessagesSettingsFragment) return@setOnClickListener
            val users = viewModel.getUsers().value
            val currentUserIds: LongArray = users?.asSequence()?.map { obj: User -> obj.pk }?.sorted()?.toList()?.toLongArray() ?: LongArray(0)
            val actionGlobalUserSearch = UserSearchFragmentDirections
                .actionGlobalUserSearch()
                .setTitle(getString(R.string.add_members))
                .setActionLabel(getString(R.string.add))
                .setHideUserIds(currentUserIds)
                .setSearchMode(UserSearchFragment.SearchMode.RAVEN)
                .setMultiple(true)
            navController.navigate(actionGlobalUserSearch)
        }
        binding.muteMentionsLabel.setOnClickListener { binding.muteMentions.toggle() }
        binding.muteMentions.setOnCheckedChangeListener { buttonView: CompoundButton, isChecked: Boolean ->
            val resourceLiveData = if (isChecked) viewModel.muteMentions() else viewModel.unmuteMentions()
            handleSwitchChangeResource(resourceLiveData, buttonView)
        }
        binding.leave.setOnClickListener {
            val confirmDialogFragment = ConfirmDialogFragment.newInstance(
                LEAVE_THREAD_REQUEST_CODE,
                R.string.dms_action_leave_question,
                0,
                R.string.yes,
                R.string.no,
                0
            )
            confirmDialogFragment.show(childFragmentManager, "leave_thread_confirmation_dialog")
        }
        var isViewerAdmin = viewModel.isViewerAdmin().value
        if (isViewerAdmin == null) isViewerAdmin = false
        if (isViewerAdmin) {
            binding.end.visibility = View.VISIBLE
            binding.end.setOnClickListener {
                val confirmDialogFragment = ConfirmDialogFragment.newInstance(
                    END_THREAD_REQUEST_CODE,
                    R.string.dms_action_end_question,
                    R.string.dms_action_end_description,
                    R.string.yes,
                    R.string.no,
                    0
                )
                confirmDialogFragment.show(childFragmentManager, "end_thread_confirmation_dialog")
            }
        } else {
            binding.end.visibility = View.GONE
        }
    }

    private fun setApprovalRelatedUI(isViewerAdmin: Boolean) {
        if (!isViewerAdmin) {
            binding.pendingMembersGroup.visibility = View.GONE
            binding.approvalRequired.visibility = View.GONE
            binding.approvalRequiredLabel.visibility = View.GONE
            return
        }
        binding.approvalRequired.visibility = View.VISIBLE
        binding.approvalRequiredLabel.visibility = View.VISIBLE
        binding.approvalRequiredLabel.setOnClickListener { binding.approvalRequired.toggle() }
        binding.approvalRequired.setOnCheckedChangeListener { buttonView: CompoundButton, isChecked: Boolean ->
            val resourceLiveData = if (isChecked) viewModel.approvalRequired() else viewModel.approvalNotRequired()
            handleSwitchChangeResource(resourceLiveData, buttonView)
        }
    }

    private fun handleSwitchChangeResource(resourceLiveData: LiveData<Resource<Any?>>, buttonView: CompoundButton) {
        resourceLiveData.observe(viewLifecycleOwner, { resource: Resource<Any?>? ->
            if (resource == null) return@observe
            when (resource.status) {
                Resource.Status.SUCCESS -> buttonView.isEnabled = true
                Resource.Status.ERROR -> {
                    buttonView.isEnabled = true
                    buttonView.isChecked = !buttonView.isChecked
                    if (resource.message != null) {
                        Snackbar.make(binding.root, resource.message, Snackbar.LENGTH_LONG).show()
                    }
                    if (resource.resId != 0) {
                        Snackbar.make(binding.root, resource.resId, Snackbar.LENGTH_LONG).show()
                    }
                }
                Resource.Status.LOADING -> buttonView.isEnabled = false
            }
        })
    }

    private fun setupMembers() {
        val context = context ?: return
        binding.users.layoutManager = LinearLayoutManager(context)
        val inviter = viewModel.getInviter().value
        usersAdapter = DirectUsersAdapter(
            inviter?.pk ?: -1,
            { _: Int, user: User, _: Boolean ->
                if (user.username.isBlank() && !user.interopMessagingUserFbid.isNullOrBlank()) {
                    Utils.openURL(context, "https://facebook.com/" + user.interopMessagingUserFbid)
                    return@DirectUsersAdapter
                }
                if (isEmpty(user.username)) return@DirectUsersAdapter
                val directions = ProfileNavGraphDirections
                    .actionGlobalProfileFragment("@" + user.username)
                NavHostFragment.findNavController(this).navigate(directions)
            },
            { _: Int, user: User? ->
                val options = viewModel.createUserOptions(user)
                if (options.isEmpty()) return@DirectUsersAdapter true
                val fragment = MultiOptionDialogFragment.newInstance(-1, options)
                fragment.setSingleCallback(object : MultiOptionDialogSingleCallback<String?> {
                    override fun onSelect(action: String?) {
                        if (action == null) return
                        val resourceLiveData = viewModel.doAction(user, action)
                        if (resourceLiveData != null) {
                            observeDetailsChange(resourceLiveData)
                        }
                    }

                    override fun onCancel() {}
                })
                val fragmentManager = childFragmentManager
                fragment.show(fragmentManager, "actions")
                true
            }
        )
        binding.users.adapter = usersAdapter
    }

    private fun setPendingRequests(requests: DirectThreadParticipantRequestsResponse?) {
        val nullOrEmpty: Boolean = requests?.users?.isNullOrEmpty() ?: true
        if (nullOrEmpty) {
            binding.pendingMembersGroup.visibility = View.GONE
            return
        }
        if (!isPendingRequestsSetupDone) {
            val context = context ?: return
            binding.pendingMembers.layoutManager = LinearLayoutManager(context)
            pendingUsersAdapter = DirectPendingUsersAdapter(object : PendingUserCallback {
                override fun onClick(position: Int, pendingUser: PendingUser) {
                    val directions = ProfileNavGraphDirections
                        .actionGlobalProfileFragment("@" + pendingUser.user.username)
                    NavHostFragment.findNavController(this@DirectMessageSettingsFragment).navigate(directions)
                }

                override fun onApprove(position: Int, pendingUser: PendingUser) {
                    val resourceLiveData = viewModel.approveUsers(listOf(pendingUser.user))
                    observeApprovalChange(resourceLiveData, position, pendingUser)
                }

                override fun onDeny(position: Int, pendingUser: PendingUser) {
                    val resourceLiveData = viewModel.denyUsers(listOf(pendingUser.user))
                    observeApprovalChange(resourceLiveData, position, pendingUser)
                }
            })
            binding.pendingMembers.adapter = pendingUsersAdapter
            binding.pendingMembersGroup.visibility = View.VISIBLE
            isPendingRequestsSetupDone = true
        }
        pendingUsersAdapter?.submitPendingRequests(requests)
    }

    private fun observeDetailsChange(resourceLiveData: LiveData<Resource<Any?>>) {
        resourceLiveData.observe(viewLifecycleOwner, { resource: Resource<Any?>? ->
            if (resource == null) return@observe
            when (resource.status) {
                Resource.Status.SUCCESS,
                Resource.Status.LOADING,
                -> {
                }
                Resource.Status.ERROR -> {
                    if (resource.message != null) {
                        Snackbar.make(binding.root, resource.message, Snackbar.LENGTH_LONG).show()
                    }
                    if (resource.resId != 0) {
                        Snackbar.make(binding.root, resource.resId, Snackbar.LENGTH_LONG).show()
                    }
                }
            }
        })
    }

    private fun observeApprovalChange(
        detailsChangeResourceLiveData: LiveData<Resource<Any?>>,
        position: Int,
        pendingUser: PendingUser,
    ) {
        detailsChangeResourceLiveData.observe(viewLifecycleOwner, { resource: Resource<Any?>? ->
            if (resource == null) return@observe
            when (resource.status) {
                Resource.Status.SUCCESS -> {
                }
                Resource.Status.LOADING -> pendingUser.isInProgress = true
                Resource.Status.ERROR -> {
                    pendingUser.isInProgress = false
                    if (resource.message != null) {
                        Snackbar.make(binding.root, resource.message, Snackbar.LENGTH_LONG).show()
                    }
                    if (resource.resId != 0) {
                        Snackbar.make(binding.root, resource.resId, Snackbar.LENGTH_LONG).show()
                    }
                }
            }
            pendingUsersAdapter?.notifyItemChanged(position)
        })
    }

    override fun onPositiveButtonClicked(requestCode: Int) {
        if (requestCode == APPROVAL_REQUIRED_REQUEST_CODE) {
            approvalRequiredUsers?.let {
                val detailsChangeResourceLiveData = viewModel.addMembers(it)
                observeDetailsChange(detailsChangeResourceLiveData)
            }
            return
        }
        if (requestCode == LEAVE_THREAD_REQUEST_CODE) {
            val resourceLiveData = viewModel.leave()
            resourceLiveData.observe(viewLifecycleOwner, { resource: Resource<Any?>? ->
                if (resource == null) return@observe
                when (resource.status) {
                    Resource.Status.SUCCESS -> {
                        val directions = DirectMessageSettingsFragmentDirections.actionSettingsToInbox()
                        NavHostFragment.findNavController(this).navigate(directions)
                    }
                    Resource.Status.ERROR -> {
                        binding.leave.isEnabled = true
                        if (resource.message != null) {
                            Snackbar.make(binding.root, resource.message, Snackbar.LENGTH_LONG).show()
                        }
                        if (resource.resId != 0) {
                            Snackbar.make(binding.root, resource.resId, Snackbar.LENGTH_LONG).show()
                        }
                    }
                    Resource.Status.LOADING -> binding.leave.isEnabled = false
                }
            })
            return
        }
        if (requestCode == END_THREAD_REQUEST_CODE) {
            val resourceLiveData = viewModel.end()
            resourceLiveData.observe(viewLifecycleOwner, { resource: Resource<Any?>? ->
                if (resource == null) return@observe
                when (resource.status) {
                    Resource.Status.SUCCESS -> {
                    }
                    Resource.Status.ERROR -> {
                        binding.end.isEnabled = true
                        if (resource.message != null) {
                            Snackbar.make(binding.root, resource.message, Snackbar.LENGTH_LONG).show()
                        }
                        if (resource.resId != 0) {
                            Snackbar.make(binding.root, resource.resId, Snackbar.LENGTH_LONG).show()
                        }
                    }
                    Resource.Status.LOADING -> binding.end.isEnabled = false
                }
            })
        }
    }

    override fun onNegativeButtonClicked(requestCode: Int) {
        if (requestCode == APPROVAL_REQUIRED_REQUEST_CODE) {
            approvalRequiredUsers = null
        }
    }

    override fun onNeutralButtonClicked(requestCode: Int) {}

    companion object {
        private const val APPROVAL_REQUIRED_REQUEST_CODE = 200
        private const val LEAVE_THREAD_REQUEST_CODE = 201
        private const val END_THREAD_REQUEST_CODE = 202
    }
}